swiftpkg/bzlmod/swift_deps.bzl

```diff
@@ -1,39 +1,154 @@
 """Implementation for `swift_deps` bzlmod extension."""
 
-load("//swiftpkg/internal:deps_indexes.bzl", "deps_indexes")
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("//swiftpkg/internal:bazel_repo_names.bzl", "bazel_repo_names")
 load("//swiftpkg/internal:local_swift_package.bzl", "local_swift_package")
-load("//swiftpkg/internal:swift_package.bzl", "swift_package")
+load("//swiftpkg/internal:pkginfos.bzl", "pkginfos")
+load("//swiftpkg/internal:swift_deps_info.bzl", "swift_deps_info")
+load("//swiftpkg/internal:swift_package.bzl", "PATCH_ATTRS", "swift_package")
 
 # MARK: - swift_deps bzlmod Extension
 
-def _declare_pkg_from_package(package, deps_index_label, config_pkg):
-    if package.remote_pkg != None:
-        remote_pkg = package.remote_pkg
+def _declare_pkgs_from_package(module_ctx, from_package, config_pkgs):
+    """Declare Swift packages from `Package.swift` and `Package.resolved`.
+
+    Args:
+        module_ctx: An instance of `module_ctx`.
+        from_package: The data from the `from_package` tag.
+        config_pkgs: The data from the `configure_package` tag.
+    """
+
+    # Read Package.resolved.
+    if from_package.resolved:
+        pkg_resolved = module_ctx.path(from_package.resolved)
+        resolved_pkg_json = module_ctx.read(pkg_resolved)
+        resolved_pkg_map = json.decode(resolved_pkg_json)
+    else:
+        resolved_pkg_map = dict()
+
+    # Get the package info.
+    pkg_swift = module_ctx.path(from_package.swift)
+    debug_path = module_ctx.path(".")
+    pkg_info = pkginfos.get(
+        module_ctx,
+        directory = str(pkg_swift.dirname),
+        debug_path = str(debug_path),
+        resolved_pkg_map = resolved_pkg_map,
+        collect_src_info = False,
+    )
+
+    # Collect all of the deps by identity
+    all_deps_by_id = {
+        dep.identity: dep
+        for dep in pkg_info.dependencies
+    }
+
+    # Collect the direct dep repo names
+    direct_dep_repo_names = []
+    direct_dep_pkg_infos = {}
+    for dep in pkg_info.dependencies:
+        bazel_repo_name = bazel_repo_names.from_identity(dep.identity)
+        direct_dep_repo_names.append(bazel_repo_name)
+        pkg_info_label = "@{}//:pkg_info.json".format(bazel_repo_name)
+        direct_dep_pkg_infos[pkg_info_label] = dep.identity
+
+    # Write info about the Swift deps that may be used by external tooling.
+    swift_deps_info_repo_name = "swift_deps_info"
+    swift_deps_info(
+        name = swift_deps_info_repo_name,
+        direct_dep_pkg_infos = direct_dep_pkg_infos,
+    )
+    direct_dep_repo_names.append(swift_deps_info_repo_name)
```

Comment:
This should be done in a different extension, so projects that don't use it won't incur the cost (or the `use_repo` warning).

Reply:
I spent some time trying to move this to a separate extension and not have to manually enter the Swift package identity to pkg_info.json label mapping. It turns out to be non-trivial. For now, I have added a `bool` attribute that controls whether the `swift_deps_info` is declared. The default is `False`.